rust/worker/src/compactor/scheduler.rs

```diff
@@ -512,56 +526,105 @@ impl Scheduler {
                     )
                     .await;
-                self.schedule_tasks(tasks);
+                self.schedule_tasks(tasks).await;
             }
         }
     }
 
-    pub(crate) fn schedule_tasks(&mut self, tasks: Vec<SchedulableTask>) {
+    pub(crate) async fn schedule_tasks(&mut self, funcs: Vec<SchedulableFunction>) {
         let members = self.memberlist.as_ref().unwrap();
         let members_as_string = members
             .iter()
             .map(|member| member.member_id.clone())
             .collect();
         self.assignment_policy.set_members(members_as_string);
-        for task in tasks {
+        for func in funcs {
             let result = self
                 .assignment_policy
-                .assign_one(task.collection_id.0.to_string().as_str());
+                .assign_one(func.collection_id.0.to_string().as_str());
             if result.is_err() {
                 tracing::error!(
-                    "Failed to assign task {} for collection {} to member: {}",
-                    task.task_id,
-                    task.collection_id,
+                    "Failed to assign func {} for collection {} to member: {}",
+                    func.task_id,
+                    func.collection_id,
                     result.err().unwrap()
                 );
                 continue;
             }
-
             let member = result.unwrap();
             if member != self.my_member_id {
                 continue;
             }
 
-            tracing::info!(
-                "SCHEDULING TASKS FOR {:?} {:?} {:?} {:?}",
-                task.bucket,
-                task.collection_id,
-                task.task_id,
-                task.nonce
-            );
+            let failure_count = self
+                .failing_jobs
+                .get(&func.collection_id.into())
+                .map(|job| job.failure_count())
+                .unwrap_or(0);
 
-            // TODO: dedup with already-enqueued tasks
-            self.task_queue.push(task);
+            if failure_count >= self.max_failure_count {
+                tracing::warn!(
+                    "Job for collection {} failed more than {} times, moving this to dead jobs and skipping function for it",
+                    func.collection_id,
+                    self.max_failure_count
+                );
+                self.kill_job(func.collection_id.into());
+                continue;
+            }
+
+            if self.disabled_collections.contains(&func.collection_id)
+                || self.dead_jobs.contains(&func.collection_id.into())
+            {
+                tracing::info!(
+                    "Ignoring collection: {:?} because it disabled",
+                    func.collection_id
+                );
+                continue;
+            }
+            if let Entry::Vacant(entry) = self.in_progress_jobs.entry(func.nonce.into()) {
```

Comment:
See `is_job_in_progress`. Might be able to slightly refactor that to get rid of the code duplication here.